<h1> House Sales Project </h1>
<h2> Approach Statement </h2>
<p>
    Approach Statement will go here
</p>

In [2]:
# First I will import modules required to explore the data.
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
# Next I will load the data into a pandas dataframe.
house_data = pd.read_csv('kc_house_data.csv')

In [4]:
# Now I will explore the data to understand what it looks like and what I am working with.
house_data.head()

,id,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,...,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
0,7129300520,10/13/2014,221900.0,3,1.00,1180,5650,1.0,NaN,0.0,...,7,1180,0.0,1955,0.0,98178,47.5112,-122.257,1340,5650
1,6414100192,12/9/2014,538000.0,3,2.25,2570,7242,2.0,0.0,0.0,...,7,2170,400.0,1951,1991.0,98125,47.7210,-122.319,1690,7639
2,5631500400,2/25/2015,180000.0,2,1.00,770,10000,1.0,0.0,0.0,...,6,770,0.0,1933,NaN,98028,47.7379,-122.233,2720,8062
3,2487200875,12/9/2014,604000.0,4,3.00,1960,5000,1.0,0.0,0.0,...,7,1050,910.0,1965,0.0,98136,47.5208,-122.393,1360,5000
4,1954400510,2/18/2015,510000.0,3,2.00,1680,8080,1.0,0.0,0.0,...,8,1680,0.0,1987,0.0,98074,47.6168,-122.045,1800,7503


In [5]:
house_data.tail()

,id,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,...,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
21592,263000018,5/21/2014,360000.0,3,2.50,1530,1131,3.0,0.0,0.0,...,8,1530,0.0,2009,0.0,98103,47.6993,-122.346,1530,1509
21593,6600060120,2/23/2015,400000.0,4,2.50,2310,5813,2.0,0.0,0.0,...,8,2310,0.0,2014,0.0,98146,47.5107,-122.362,1830,7200
21594,1523300141,6/23/2014,402101.0,2,0.75,1020,1350,2.0,0.0,0.0,...,7,1020,0.0,2009,0.0,98144,47.5944,-122.299,1020,2007
21595,291310100,1/16/2015,400000.0,3,2.50,1600,2388,2.0,NaN,0.0,...,8,1600,0.0,2004,0.0,98027,47.5345,-122.069,1410,1287
21596,1523300157,10/15/2014,325000.0,2,0.75,1020,1076,2.0,0.0,0.0,...,7,1020,0.0,2008,0.0,98144,47.5941,-122.299,1020,1357


In [6]:
house_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21597 non-null  int64  
 1   date           21597 non-null  str    
 2   price          21597 non-null  float64
 3   bedrooms       21597 non-null  int64  
 4   bathrooms      21597 non-null  float64
 5   sqft_living    21597 non-null  int64  
 6   sqft_lot       21597 non-null  int64  
 7   floors         21597 non-null  float64
 8   waterfront     19221 non-null  float64
 9   view           21534 non-null  float64
 10  condition      21597 non-null  int64  
 11  grade          21597 non-null  int64  
 12  sqft_above     21597 non-null  int64  
 13  sqft_basement  21597 non-null  str    
 14  yr_built       21597 non-null  int64  
 15  yr_renovated   17755 non-null  float64
 16  zipcode        21597 non-null  int64  
 17  lat            21597 non-null  float64
 18  long           21

In [7]:
house_data.describe()

,id,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,condition,grade,sqft_above,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
count,2.159700e+04,2.159700e+04,21597.000000,21597.000000,21597.000000,2.159700e+04,21597.000000,19221.000000,21534.000000,21597.000000,21597.000000,21597.000000,21597.000000,17755.000000,21597.000000,21597.000000,21597.000000,21597.000000,21597.000000
mean,4.580474e+09,5.402966e+05,3.373200,2.115826,2080.321850,1.509941e+04,1.494096,0.007596,0.233863,3.409825,7.657915,1788.596842,1970.999676,83.636778,98077.951845,47.560093,-122.213982,1986.620318,12758.283512
std,2.876736e+09,3.673681e+05,0.926299,0.768984,918.106125,4.141264e+04,0.539683,0.086825,0.765686,0.650546,1.173200,827.759761,29.375234,399.946414,53.513072,0.138552,0.140724,685.230472,27274.441950
min,1.000102e+06,7.800000e+04,1.000000,0.500000,370.000000,5.200000e+02,1.000000,0.000000,0.000000,1.000000,3.000000,370.000000,1900.000000,0.000000,98001.000000,47.155900,-122.519000,399.000000,651.000000
25%,2.123049e+09,3.220000e+05,3.000000,1.750000,1430.000000,5.040000e+03,1.000000,0.000000,0.000000,3.000000,7.000000,1190.000000,1951.000000,0.000000,98033.000000,47.471100,-122.328000,1490.000000,5100.000000
50%,3.904930e+09,4.500000e+05,3.000000,2.250000,1910.000000,7.618000e+03,1.500000,0.000000,0.000000,3.000000,7.000000,1560.000000,1975.000000,0.000000,98065.000000,47.571800,-122.231000,1840.000000,7620.000000
75%,7.308900e+09,6.450000e+05,4.000000,2.500000,2550.000000,1.068500e+04,2.000000,0.000000,0.000000,4.000000,8.000000,2210.000000,1997.000000,0.000000,98118.000000,47.678000,-122.125000,2360.000000,10083.000000
max,9.900000e+09,7.700000e+06,33.000000,8.000000,13540.000000,1.651359e+06,3.500000,1.000000,4.000000,5.000000,13.000000,9410.000000,2015.000000,2015.000000,98199.000000,47.777600,-121.315000,6210.000000,871200.000000


<h2> Initial Findings </h2>
<ul>
    <li>There are 21597 rows.</li>
    <li>Date is stored as str, this may need converting.</li>
    <li>sqft_basement is stored as str, head values appear to be float? Will explore.</li>
    <li>The following columns contain null values which I will need to review:
        <ul>
            <li>waterfront</li>
            <li>view</li>
            <li>yr_renovated</li>
        </ul>
    </li>
    <li> Max bedrooms of 33 seems unusually high</li>
    <li>Zipcode is numerical, should this be str? Better as categorical data</li>
</ul>

<h2> Data Cleaning </h2>

<h3> Convert date column from str to datetime</h3>

In [8]:
# I will fix the date column to be in a proper datetime format "dd/mm/yyyy".
house_data['date'] = pd.to_datetime(house_data['date'])
house_data['date']

0       2014-10-13
1       2014-12-09
2       2015-02-25
3       2014-12-09
4       2015-02-18
           ...    
21592   2014-05-21
21593   2015-02-23
21594   2014-06-23
21595   2015-01-16
21596   2014-10-15
Name: date, Length: 21597, dtype: datetime64[us]

<h3>Review the sqft_basement column and decide format</h3>

In [9]:
# I will check sqft_basement as showing data stored as string instead of int.
house_data['sqft_basement'].unique()

<StringArray>
[   '0.0',  '400.0',  '910.0', '1530.0',      '?',  '730.0', '1700.0',
  '300.0',  '970.0',  '760.0',
 ...
  '207.0',  '915.0',  '556.0',  '417.0',  '143.0',  '508.0', '2810.0',
   '20.0',  '274.0',  '248.0']
Length: 304, dtype: str

In [10]:
# flag non numeric numbers from the unique list
non_numeric = [x for x in house_data['sqft_basement'].unique() 
               if pd.isna(pd.to_numeric(x, errors='coerce'))]
non_numeric

['?']

In [11]:
# Having identified the non-numeric values, I will replace them with NaN and then convert the column to numeric.
house_data['sqft_basement'] = pd.to_numeric(house_data['sqft_basement'], errors='coerce')
house_data['sqft_basement'].info()

<class 'pandas.Series'>
RangeIndex: 21597 entries, 0 to 21596
Series name: sqft_basement
Non-Null Count  Dtype  
--------------  -----  
21143 non-null  float64
dtypes: float64(1)
memory usage: 168.9 KB


<h3>Clean columns with NaN values</h3>

The following columns contain Nan values
<ul>
    <li>waterfront</li>
    <li>view</li>
    <li>yr_renovated</li>
    <li>sqft_basement</li>
</ul>

<h4>--> waterfront</h4>

In [12]:
# First lets check waterfront and assess missing values
house_data['waterfront'].unique()

array([nan,  0.,  1.])

In [25]:
# How many missing values are there in the waterfront column?
house_data['waterfront'].isna().sum() 

np.int64(2376)

In [20]:
# The waterfront column has 0, 1 and NaN values. I will check the percentage of missing values in this column.
house_data['waterfront'].isna().sum() / len(house_data) * 100

np.float64(11.00152798999861)

In [24]:
# I will check how many houses have a waterfront view and how many do not.
house_data['waterfront'].value_counts()

waterfront
0.0    19075
1.0      146
Name: count, dtype: int64

<b>Decision</b>
<p>
    Given that 
    <ul>
        <li>There are 2376 missing values (11%)</li>
        <li>Only 146 known values have a waterfront (less than 1%)</li>
    </ul>
     I will treat all NaN as 0.
</p>

In [27]:
# Convert waterfront missing values to 0, as it is likely that the missing values indicate no waterfront view.
house_data['waterfront'] = house_data['waterfront'].fillna(0)
house_data['waterfront'].value_counts()

waterfront
0.0    21451
1.0      146
Name: count, dtype: int64

<h4>--> view</h4>

In [28]:
# First lets check view and assess missing values
house_data['view'].unique()

array([ 0., nan,  3.,  4.,  2.,  1.])

In [33]:
# How many missing values are there in the view column?
house_data['view'].isna().sum() 

np.int64(63)

In [34]:
# There are 63 missing values in the view column. I will check the percentage of missing values in this column.
house_data['view'].isna().sum() / len(house_data) * 100

np.float64(0.29170718155299347)

In [36]:
# I will check for how values are currently distributed in the view column.
house_data['view'].value_counts()

view
0.0    19422
2.0      957
3.0      508
1.0      330
4.0      317
Name: count, dtype: int64

In [40]:
# Mean mode and median of the view column to determine which is the best approach to filling the missing values.
house_data['view'].mean(), house_data['view'].mode()[0], house_data['view'].median()

(np.float64(0.23386272870808952), np.float64(0.0), np.float64(0.0))

<b>Decision</b>
<p>
    Given that 
    <ul>
        <li>There are 63 missing values (less than 1%)</li>
        <li>View are in ordinal values, 0 the most common</li>
        <li>Median and Mode = 0</li>
        <li>Mean is close to 0</li>
    </ul>
     I will treat all NaN as 0.
</p>

In [43]:
# convert the missing values in the view column to 0, as it is likely that the missing values indicate no view.
house_data['view'] = house_data['view'].fillna(0)
house_data['view'].value_counts().sort_index()

view
0.0    19485
1.0      330
2.0      957
3.0      508
4.0      317
Name: count, dtype: int64

<h4>--> yr_renovated</h4>

In [45]:
# First lets check yr_renovated and assess missing values
house_data['yr_renovated'].unique()

array([   0., 1991.,   nan, 2002., 2010., 1992., 2013., 1994., 1978.,
       2005., 2003., 1984., 1954., 2014., 2011., 1983., 1945., 1990.,
       1988., 1977., 1981., 1995., 2000., 1999., 1998., 1970., 1989.,
       2004., 1986., 2007., 1987., 2006., 1985., 2001., 1980., 1971.,
       1979., 1997., 1950., 1969., 1948., 2009., 2015., 1974., 2008.,
       1968., 2012., 1963., 1951., 1962., 1953., 1993., 1996., 1955.,
       1982., 1956., 1940., 1976., 1946., 1975., 1964., 1973., 1957.,
       1959., 1960., 1967., 1965., 1934., 1972., 1944., 1958.])

In [47]:
# How many missing values are there in the yr_renovated column?
house_data['yr_renovated'].isna().sum() 

np.int64(3842)

In [54]:
# There are 3842 missing values in the yr_renovated column. I will check the percentage of missing values in this column.
house_data['yr_renovated'].isna().sum() / len(house_data) * 100

np.float64(17.78950780200954)

In [51]:
# Check the value counts of the yr_renovated column how these are distributed.
house_data['yr_renovated'].value_counts()

yr_renovated
0.0       17011
2014.0       73
2013.0       31
2003.0       31
2007.0       30
          ...  
1976.0        1
1946.0        1
1959.0        1
1934.0        1
1944.0        1
Name: count, Length: 70, dtype: int64

In [53]:
# Mode and median of the yr_renovated column to determine which is the best approach to filling the missing values.
house_data['yr_renovated'].mode()[0], house_data['yr_renovated'].median()

(np.float64(0.0), np.float64(0.0))

<b>Decision</b>
<p>
    Given that
    <ul>
        <li>There are 3842 missing values (18%)</li>
        <li>17,011 rows have a value of 0, meaning never renovated</li>
        <li>NaN likely also means never renovated</li>
        <li>Knowing whether the house has been renovated will be more important than when</li>
    </ul>
    I will create a new binary column "renovated" (1 = renovated, 0 = not renovated),
    treating both 0 and NaN as not renovated. The yr_renovated column will be dropped.
</p>

In [55]:
# New binary column to indicate whether a house has been renovated or not.
house_data['renovated'] = house_data['yr_renovated'].apply(lambda x: 1 if x > 0 else 0)
house_data['renovated'].value_counts()

renovated
0    20853
1      744
Name: count, dtype: int64

In [60]:
# Now i will drop the yr_renovated column as it is no longer needed.
house_data = house_data.drop('yr_renovated', axis=1, errors='ignore')
house_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   id             21597 non-null  int64         
 1   date           21597 non-null  datetime64[us]
 2   price          21597 non-null  float64       
 3   bedrooms       21597 non-null  int64         
 4   bathrooms      21597 non-null  float64       
 5   sqft_living    21597 non-null  int64         
 6   sqft_lot       21597 non-null  int64         
 7   floors         21597 non-null  float64       
 8   waterfront     21597 non-null  float64       
 9   view           21597 non-null  float64       
 10  condition      21597 non-null  int64         
 11  grade          21597 non-null  int64         
 12  sqft_above     21597 non-null  int64         
 13  sqft_basement  21143 non-null  float64       
 14  yr_built       21597 non-null  int64         
 15  zipcode        21597 non-null 

<h4>--> sqft_basement</h4>

In [61]:
# First lets check sqft_basement and assess missing values
house_data['sqft_basement'].unique()

array([   0.,  400.,  910., 1530.,   nan,  730., 1700.,  300.,  970.,
        760.,  720.,  700.,  820.,  780.,  790.,  330., 1620.,  360.,
        588., 1510.,  410.,  990.,  600.,  560.,  550., 1000., 1600.,
        500., 1040.,  880., 1010.,  240.,  265.,  290.,  800.,  540.,
        710.,  840.,  380.,  770.,  480.,  570., 1490.,  620., 1250.,
       1270.,  120.,  650.,  180., 1130.,  450., 1640., 1460., 1020.,
       1030.,  750.,  640., 1070.,  490., 1310.,  630., 2000.,  390.,
        430.,  850.,  210., 1430., 1950.,  440.,  220., 1160.,  860.,
        580., 2060., 1820., 1180.,  200., 1150., 1200.,  680.,  530.,
       1450., 1170., 1080.,  960.,  280.,  870., 1100.,  460., 1400.,
        660., 1220.,  900.,  420., 1580., 1380.,  475.,  690.,  270.,
        350.,  935., 1370.,  980., 1470.,  160.,  950.,   50.,  740.,
       1780., 1900.,  340.,  470.,  370.,  140., 1760.,  130.,  520.,
        890., 1110.,  150., 1720.,  810.,  190., 1290.,  670., 1800.,
       1120., 1810.,

In [62]:
# How many missing values are there in the sqft_basement column?
house_data['sqft_basement'].isna().sum() 

np.int64(454)

In [63]:
# Percentage of missing values in the sqft_basement column.
house_data['sqft_basement'].isna().sum() / len(house_data) * 100

np.float64(2.1021438162707784)

In [64]:
# Check the value counts of the sqft_basement column how these are distributed.
house_data['sqft_basement'].value_counts()

sqft_basement
0.0       12826
600.0       217
500.0       209
700.0       208
800.0       201
          ...  
508.0         1
2810.0        1
20.0          1
274.0         1
248.0         1
Name: count, Length: 303, dtype: int64

In [65]:
# Mode and median of the sqft_basement column to determine which is the best approach to filling the missing values.
house_data['sqft_basement'].mode()[0], house_data['sqft_basement'].median()

(np.float64(0.0), np.float64(0.0))

There is a different approach here, there is additional information relating to sqft that might provide the missing values....

In [69]:
# Check if sqft_living - sqft_above = sqft_basement for non-null rows
house_data['sqft_derived'] = house_data['sqft_living'] - house_data['sqft_above']

# Compare derived vs actual on non-null rows
comparison = house_data[house_data['sqft_basement'].notna()][['sqft_basement', 'sqft_derived']]
comparison.head(10)

,sqft_basement,sqft_derived
0,0.0,0
1,400.0,400
2,0.0,0
3,910.0,910
4,0.0,0
5,1530.0,1530
7,0.0,0
8,730.0,730
9,0.0,0
10,1700.0,1700


In [70]:
# Check how many rows match exactly
exact_match = (comparison['sqft_basement'] == comparison['sqft_derived']).sum()
total = len(comparison)
print(f"Exact matches: {exact_match} / {total} ({exact_match/total*100:.1f}%)")

Exact matches: 21143 / 21143 (100.0%)


<b>Decision</b>
<p>
    Given that
    <ul>
        <li>There are NaN values in the sqft_basement column</li>
        <li>The dataset contains sqft_living (total interior space) and sqft_above (above ground space)</li>
        <li>A 100% exact match was confirmed between sqft_basement and sqft_living - sqft_above for all non-null rows</li>
    </ul>
    The missing values can be derived exactly using:
     sqft_basement = sqft_living - sqft_above.
    This is a mathematically exact relationship, not an assumption.
</p>

In [71]:
# 100% match confirmed: sqft_basement = sqft_living - sqft_above
# Fill NaN values using this derived relationship
house_data['sqft_basement'] = house_data['sqft_basement'].fillna(house_data['sqft_derived'])

# Drop the temporary derived column
house_data = house_data.drop('sqft_derived', axis=1)

house_data['sqft_basement'].isna().sum() 

np.int64(0)

<h3>Review the max 33 Bedrroms</h3>

<h3>Review zipcode</h3>